### PR TITLE
Add ArtifactVersion to KubernetesApp Deployment

### PR DESCRIPTION
**Pull request by ffjlabo**
**What this PR does / why we need it**:

Add rich version info to KubernetesApp Deployment

**Which issue(s) this PR fixes**:

A part of #3303 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/app/piped/planner/kubernetes/kubernetes.go [hunk 1]
```diff
@@ -97,6 +97,18 @@ func (p *Planner) Plan(ctx context.Context, in planner.Input) (out planner.Outpu
 		out.Version = version
 	}
 
+	if versions, e := determineVersions(newManifests); e != nil {
+		in.Logger.Error("unable to determine versions", zap.Error(e))
+		out.Versions = []*model.ArtifactVersion{
+			{
+				Kind:    model.ArtifactVersion_CONTAINER_IMAGE,
```
Comment (nghialv) [inline pkg/app/piped/planner/kubernetes/kubernetes.go:104]: nit: We can use this kind https://github.com/pipe-cd/pipecd/blob/0c6a1b3f464b659e83cc5f1aa280fa8f914e4c92/pkg/model/application.proto#L90

File: pkg/app/piped/planner/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -480,3 +492,45 @@ func determineVersion(manifests []provider.Manifest) (string, error) {
 
 	return b.String(), nil
 }
+
+// determineVersions decide artifact versions of an application.
+// For example, in case some containers are used in one application, those image version are retured
```
Comment (nghialv) [inline pkg/app/piped/planner/kubernetes/kubernetes.go:497]:
```suggestion
// determineVersions decides artifact versions of an application.
// It finds all container images that are being specified in the workload manifests then returns their names, version numbers, and urls.
```

File: pkg/app/piped/planner/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -480,3 +492,45 @@ func determineVersion(manifests []provider.Manifest) (string, error) {
...
+		for _, c := range containers {
+			image := parseContainerImage(c.Image)
+			versions = append(versions, &model.ArtifactVersion{
+				Kind:    model.ArtifactVersion_CONTAINER_IMAGE,
+				Version: image.tag,
+				Name:    image.name,
+				Url:     c.Image,
+			})
+		}
+	}
```
Comment (nghialv) [inline pkg/app/piped/planner/kubernetes/kubernetes.go:524]: Same container images could be used in multiple manifests so I think we have to remove the duplicate ones.

**Comment by nghialv** [comment]
@ffjlabo Great work! Love to see this.

I have just left some comments. Please check them.

File: pkg/model/deployment.proto [hunk 4]
```diff
@@ -146,3 +147,15 @@ message Commit {
...
+    enum Kind {
+        UNKNOWN = 0;
+        CONTAINER_IMAGE = 1;
+    }
+   
+    Kind kind = 1 [(validate.rules).enum.defined_only = true];
+    string version = 2 [(validate.rules).string.min_len = 1];
+    string name = 3;
+    string url = 4;
+}
```
Comment (khanhtc1202) [inline pkg/model/deployment.proto:161]: if we use it in both application and deployment proto, should we move it to common.proto?

Comment (ffjlabo) [reply pkg/model/deployment.proto:161]: sounds nice! I'll fix it

File: pkg/model/application.proto [hunk 5]
```diff
@@ -85,25 +85,13 @@ message ApplicationSyncState {
...
-}
-
 message ApplicationDeploymentReference {
     string deployment_id = 1 [(validate.rules).string.min_len = 1];
     DeploymentTrigger trigger = 2 [(validate.rules).message.required = true];
     string summary = 3;
     string version = 4;
     string config_filename = 5;
-    repeated ArtifactVersion versions = 6;
+    repeated model.ArtifactVersion versions = 6;
```
Comment (nghialv) [inline pkg/model/application.proto:94]:
```suggestion
    repeated ArtifactVersion versions = 6;
```

File: pkg/app/piped/planner/kubernetes/kubernetes.go [hunk 6]
```diff
@@ -480,3 +492,51 @@ func determineVersion(manifests []provider.Manifest) (string, error) {
 
 	return b.String(), nil
 }
+
+// determineVersions decides artifact versions of an application.
+// It finds all container images that are being specified in the workload manifests then returns their names, version numbers, and urls.
+func determineVersions(manifests []provider.Manifest) ([]*model.ArtifactVersion, error) {
+	versions := []*model.ArtifactVersion{}
+	imageMap := map[string]string{}
```
Comment (nghialv) [inline pkg/app/piped/planner/kubernetes/kubernetes.go:500]: To have smaller size.

```suggestion
	imageMap := map[string]struct{}{}
```

File: pkg/app/piped/planner/kubernetes/kubernetes.go [hunk 7]
```diff
@@ -480,3 +492,51 @@ func determineVersion(manifests []provider.Manifest) (string, error) {
...
+		}
+		var d resource.Deployment
+		if err := json.Unmarshal(data, &d); err != nil {
+			return nil, err
+		}
+
+		containers := d.Spec.Template.Spec.Containers
+		// remove duplicate images on multiple manifests
+		for _, c := range containers {
+			imageMap[c.Image] = ""
```
Comment (nghialv) [inline pkg/app/piped/planner/kubernetes/kubernetes.go:518]:
```suggestion
			imageMap[c.Image] = struct{}
```

File: pkg/app/piped/planner/kubernetes/kubernetes.go [hunk 8]
```diff
@@ -480,3 +492,51 @@ func determineVersion(manifests []provider.Manifest) (string, error) {
...
 }
+
+// determineVersions decides artifact versions of an application.
+// It finds all container images that are being specified in the workload manifests then returns their names, version numbers, and urls.
+func determineVersions(manifests []provider.Manifest) ([]*model.ArtifactVersion, error) {
+	versions := []*model.ArtifactVersion{}
+	imageMap := map[string]string{}
+
+	for _, m := range manifests {
+		if !m.Key.IsDeployment() {
```
Comment (nghialv) [inline pkg/app/piped/planner/kubernetes/kubernetes.go:503]:
```suggestion
                 // TODO: Determine container image version form other workload kinds such as StatefulSet, Pod, Daemon, CronJob...
		if !m.Key.IsDeployment() {
```

File: pkg/app/piped/planner/kubernetes/kubernetes.go [hunk 9]
```diff
@@ -480,3 +492,51 @@ func determineVersion(manifests []provider.Manifest) (string, error) {
...
+	}
+
+	if len(versions) == 0 {
+		return []*model.ArtifactVersion{
+			{
+				Kind:    model.ArtifactVersion_UNKNOWN,
+				Version: versionUnknown,
+			},
+		}, nil
+	}
```
Comment (nghialv) [inline pkg/app/piped/planner/kubernetes/kubernetes.go:539]: We don't need this part. An empty list is fine. It means that this application is not having any artifact version.

File: pkg/model/deployment.proto [hunk 10]
```diff
@@ -77,6 +77,7 @@ message Deployment {
     // e.g. Update image from v1.5.0 to v1.6.0.
     string summary = 22;
     string version = 23;
+    repeated ArtifactVersion versions = 103;
```
Comment (nghialv) [inline pkg/model/deployment.proto:80]:
```suggestion
    repeated ArtifactVersion versions = 24;
```

File: pkg/app/piped/planner/kubernetes/kubernetes.go [hunk 11]
```diff
@@ -480,3 +492,43 @@ func determineVersion(manifests []provider.Manifest) (string, error) {
...
+			return nil, err
+		}
+
+		containers := d.Spec.Template.Spec.Containers
+		// remove duplicate images on multiple manifests
+		for _, c := range containers {
+			imageMap[c.Image] = struct{}{}
+		}
+	}
+
```
Comment (nghialv) [inline pkg/app/piped/planner/kubernetes/kubernetes.go:522]: Move initialization versions slice to later to know its size.

```suggestion
	imageMap := map[string]struct{}{}

	for _, m := range manifests {
		// TODO: Determine container image version from other workload kinds such as StatefulSet, Pod, Daemon, CronJob...
		if !m.Key.IsDeployment() {
			continue
		}
		data, err := m.MarshalJSON()
		if err != nil {
			return nil, err
		}
		var d resource.Deployment
		if err := json.Unmarshal(data, &d); err != nil {
			return nil, err
		}

		containers := d.Spec.Template.Spec.Containers
		// remove duplicate images on multiple manifests
		for _, c := range containers {
			imageMap[c.Image] = struct{}{}
		}
	}
	
	versions := make([]*model.ArtifactVersion, 0, len(imageMap))
```

**Comment by nghialv** [comment]
Look beautiful.
Just 2 nits.

File: pkg/app/piped/planner/kubernetes/kubernetes.go (hunk 11 above)
Comment (ffjlabo) [reply pkg/app/piped/planner/kubernetes/kubernetes.go:522]: thank you for suggestion!

File: pkg/app/piped/planner/kubernetes/kubernetes.go [hunk 12]
```diff
@@ -480,3 +492,42 @@ func determineVersion(manifests []provider.Manifest) (string, error) {
...
+		if err != nil {
+			return nil, err
+		}
+		var d resource.Deployment
+		if err := json.Unmarshal(data, &d); err != nil {
+			return nil, err
+		}
+
+		containers := d.Spec.Template.Spec.Containers
+		// remove duplicate images on multiple manifests
```
Comment (khanhtc1202) [inline pkg/app/piped/planner/kubernetes/kubernetes.go:515]:
```suggestion
		// Remove duplicate images on multiple manifests.
```
As our convention, comments start with a capital character and finish with a dot.

File: pkg/model/deployment.proto [hunk 13]
```diff
@@ -77,6 +77,7 @@ message Deployment {
     // e.g. Update image from v1.5.0 to v1.6.0.
     string summary = 22;
     string version = 23;
```
Comment (khanhtc1202) [inline pkg/model/deployment.proto:79]: not really related to this PR but please add and TODO to remind remove this version from the deployment object later
> TODO: Remove version from deployment model.

**Comment by khanhtc1202** [comment]
Nice 👍 
/lgtm

**Comment by nghialv** [comment]
Great work.
/lgtm

**Pull request merged**